### PR TITLE
stm32: Add support for ETH RMII peripheral

### PR DESCRIPTION
**Pull request by dpgeorge**
This PR is based on #3671.  It adapts that PR to work with current master.

NUCLEO_F767ZI and STM32F7DISC boards have Ethernet enabled.

### CODE REVIEW

**Comment by boochow** [comment]
This works fine on my Nucleo-F767ZI. Thank you a lot for integrating codes.
I believe that NUCLEO-F746ZG and NUCLEO-H743ZI also could run this same code when pin definitions file is provided since they have same RMII interface and PHY chip (LAN8742A).

**Comment by ziadelmolla** [comment]
Hello boochow, i would like to ask you how did you test the Ethernet  , as i am struggling to work with this topic ,would appreciate any help

**Comment by dpgeorge** [comment]
> i would like to ask you how did you test the Ethernet

Please see also #3878

**Comment by boochow** [comment]
@ziadelmolla to use this PR we have to cherry-pick a bug fix of modnetwork.c.
@dpgeorge nic.ifconfig() shows a wrong dns server address. in modnetwork.c,  at line 136, we should use 'dns' instead of '&dns'.

building this PR for Nucleo-F767ZI:
```
git clone http://github.com/micropython/micropython.git
cd micropython/
git fetch origin pull/3808/head:local-branch-name
git checkout local-branch-name
git cherry-pick 309fe39dbb14b1f715ea09c4b9de235a099c01b0
git submodule update --init
make -C mpy-cross
make -C ports/stm32 MICROPY_HW_ENABLE_ETH_RMII=1 BOARD=NUCLEO_F767ZI
make -C ports/stm32 MICROPY_HW_ENABLE_ETH_RMII=1 BOARD=NUCLEO_F767ZI deploy-stlink
```
testing the ethernet and its results:
```
>>> import network
>>> nic = network.Ethernet()
>>> nic.active(True)
>>> nic.ifconfig('dhcp')
>>> print(nic.ifconfig())
('192.168.0.105', '255.255.255.0', '192.168.0.1', '192.168.0.1')
>>> import socket
>>> socket.getaddrinfo('micropython.org', 80)
[(2, 1, 0, '', ('176.58.119.26', 80))]
>>>
```

**Comment by dpgeorge** [comment]
> nic.ifconfig() shows a wrong dns server address. in modnetwork.c, at line 136, we should use 'dns' instead of '&dns'.

Thanks!  I've now fixed this in 6fc84a74542b814a90e2f34bdd927156abf9f662

**Comment by Abdelrahman-Elnaggar** [comment]
Hello @boochow  i tried to use the steps you mentioned above on my board NUCLEO_F429ZI 
but when i try to get the firmware with this command:
make -C ports/stm32 MICROPY_HW_ENABLE_ETH_RMII=1 BOARD=NUCLEO_F429ZI 
i get this error

error2

P.S : i already made changes to the board configuration files exactly like you made to your board.

**Comment by ziadelmolla** [comment]
@dpgeorge @boochow  Thank u for your help. Everything is now working properly.

**Comment by forester3** [comment]
Should I test this PR Code on STM32F746G-disco board?

**Comment by volodink** [comment]
Hi everyone! 
Thanks to all contributed into ethernet on stm32.

Today i've ported ethernet feature (RMII) from STM32F7 to STM32F4DISC board.
I´m using  STM32F4DISC with DM-STF4BB addon board to have out of the box ethernet experience.
LAN8720A and LAN8742A work perfect (they are direct pin compatible).
I've managed to run Telnet StarWars example running via serial REPL.

In order to test network performance i used _iperf_ utility.
_iperf -c board_ip -p 8080 -t 1000 -i 5_

uPy code on the board (_main.py_):

```
import network
import socket

print('Ethernet init ...')
eth = network.Ethernet()
print('Ethernet init ... done')

eth.active(1)
eth.ifconfig('dhcp')

print(eth.ifconfig())

addr = socket.getaddrinfo(eth.ifconfig()[0], 8080)[0][-1]
s = socket.socket()
s.bind(addr)
s.listen(1)

while True:
    cl, addr = s.accept()
    print('client connected from', addr)
    cl_file = cl.makefile('rwb', 0)
    while True:
        line = cl_file.read(1024)
        if not line or line == b'\r\n':
            break
    cl.close()
```
and it showed appoximately **4.19-4.3MBit** 
**BUT**
after i **ONLY** just connected and didnt open any consoles or putty to use REPL, uUSB cable to the stm32f407 board port the speed bumped up to **6.5-7.9MBit**. 
and back if one pulls usb out of the port. 
Very very wierd behavior indeed. 

Any suggestions?
Thanks.

**Comment by rockyzhaohui** [comment]
Hi @boochow，there is a error this PR：AttributeError:`module` object has no attribute'active' ，when I test these code:

import network
import socket
eth = network.Ethernet()
eth.active(True)
eth.ifconfig('dhcp')

please help me，thank you.

**Comment by bilbolodz** [comment]
I'm trying to run micropython on STM32F746G-DISCO. Using "above recipe" ethernet is working fine but SD card is now working: (PYB: can't mount SD card message during startup). Has anyone working SD card on theses board? And second question: are they any chances to include these patch into "main branch" of the project?

**Comment by dpgeorge** [comment]
This is superseded by #4541

**Comment by rolandvs** [comment]
👍👍

File: ports/stm32/Makefile [hunk 1]
```diff
@@ -258,6 +258,8 @@ SRC_C = \
 	servo.c \
 	dac.c \
 	adc.c \
+	ethernetif.c \
+	network_ethernet.c \
```
Comment (kcoelho-or) [inline ports/stm32/Makefile]: Where did you obtain this file from?

Comment (kcoelho-or) [reply ports/stm32/Makefile]: Specifically, the network_ethernet.c file

Comment (dpgeorge) [reply ports/stm32/Makefile]: As mentioned above, this PR was superseded by #4541 and `network_ethernet.c` was replaced by `network_lan.c`

**Pull request closed without merging**